### PR TITLE
HPCC-24069 Fix changeToWrite transaction bug

### PR DESCRIPTION
**Pull request by jakesmith**
Do not hold the SDS transaction mutex (dataRWLock) whilst trying
to establish a write lock in the RTM_DELETE_ON_DISCONNECT logic.
    
Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by jakesmith** [comment]
@ghalliday - please review

File: dali/base/dasds.cpp [hunk 1]
```diff
@@ -267,7 +272,12 @@ class CLCLockBlock : public CInterface
     }
     void changeToWrite(unsigned timeout)
     {
-        lock.changeToWrite(timeout);
+        /* NB: this intentionally clears the lock, before re-establishing a write lock
+        * because we do not want to keep this read lock whilst waiting for a write,
+        * as other write transactions would pile up.
```
Comment (richardkchapman) [inline dali/base/dasds.cpp:277]: And we would deadlock...

File: dali/base/dasds.cpp [hunk 2]
```diff
@@ -7848,43 +7858,63 @@ void CCovenSDSManager::disconnect(ConnectionId id, bool deleteRoot, CLCLockBlock
...
-    if (!deleteRoot && unlock(tree->queryServerId(), id, true)) // unlock returns true if last unlock and there was a setDROLR on it
+    bool last = !deleteRoot && unlock(tree->queryServerId(), id, true); // unlock returns true if last unlock and there was a setDROLR on it
+    if (last)
         deleteRoot = true;
     if (deleteRoot)
     {
-        if (lockBlock)
+        if (lockBlock) // NB: only passed if might need read->write
+        {
             lockBlock->changeToWrite(readWriteTimeout);
```
Comment (richardkchapman) [inline dali/base/dasds.cpp:7868]: This makes me nervous!

**Comment by ghalliday** [comment]
@richardkchapman it looks correct as far as can see.

**Comment by jakesmith** [comment]
@ghalliday @richardkchapman - I've spotted an issue, don't merge yet.

**Comment by jakesmith** [comment]
@ghalliday - please re-review

File: dali/base/dasds.cpp [hunk 3]
```diff
@@ -270,18 +262,28 @@ class CLCLockBlock : public CInterface
...
-    void changeToWrite(unsigned timeout)
-    {
-        /* NB: this intentionally clears the lock, before re-establishing a write lock
-        * because we do not want to keep this read lock whilst waiting for a write,
-        * as other write transactions would pile up.
-        */
-        lock.unlock();
-        lock.lockWrite(timeout);
-    }
+    void changeToWrite(unsigned timeout);
```
Comment (richardkchapman) [inline dali/base/dasds.cpp:265]: Should this have same prototype as above?

Comment (jakesmith) [reply dali/base/dasds.cpp:265]: yes, it's in conditional code which isn't on unless a #define is changed, but should be changed.

**Comment by jakesmith** [comment]
@richardkchapman @ghalliday - left as a few commits for now, please review.

**Comment by ghalliday** [comment]
@richardkchapman I haven't seen any issues.

**Comment by jakesmith** [comment]
@richardkchapman - squashed.

**Pull request merged**